url86119f019a_solzy_hive_parse_tgz-pJ8-org_apache_hadoop_hive_ql_parse_MainJ8/src/org/apache/hadoop/hive/ql/parse/HiveParser_IdentifiersParser.java

```diff
@@ -0,0 +1,12423 @@
+			// 94:31: -> ^( TOK_HAVING havingCondition )
+			{
+				// IdentifiersParser.g:94:34: ^( TOK_HAVING havingCondition )
+				{
+				CommonTree root_1 = (CommonTree)adaptor.nil();
+				root_1 = (CommonTree)adaptor.becomeRoot((CommonTree)adaptor.create(TOK_HAVING, "TOK_HAVING"), root_1);
+				adaptor.addChild(root_1, stream_havingCondition.nextTree());
+				adaptor.addChild(root_0, root_1);
+				}
+
+			}
+
+
+			retval.tree = root_0;
+			}
+
+			}
+
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "havingClause"
+
+
+	public static class havingCondition_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "havingCondition"
+	// IdentifiersParser.g:97:1: havingCondition : expression ;
+	public final HiveParser_IdentifiersParser.havingCondition_return havingCondition() throws RecognitionException {
+		HiveParser_IdentifiersParser.havingCondition_return retval = new HiveParser_IdentifiersParser.havingCondition_return();
+		retval.start = input.LT(1);
+
+		CommonTree root_0 = null;
+
+		ParserRuleReturnScope expression25 =null;
+
+
+		 gParent.msgs.push("having condition"); 
+		try {
+			// IdentifiersParser.g:100:5: ( expression )
+			// IdentifiersParser.g:101:5: expression
+			{
+			root_0 = (CommonTree)adaptor.nil();
+
+
+			pushFollow(FOLLOW_expression_in_havingCondition422);
+			expression25=expression();
+			state._fsp--;
+			if (state.failed) return retval;
+			if ( state.backtracking==0 ) adaptor.addChild(root_0, expression25.getTree());
+
+			}
+
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "havingCondition"
+
+
+	public static class orderByClause_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "orderByClause"
+	// IdentifiersParser.g:105:1: orderByClause : ( KW_ORDER KW_BY LPAREN columnRefOrder ( COMMA columnRefOrder )* RPAREN -> ^( TOK_ORDERBY ( columnRefOrder )+ ) | KW_ORDER KW_BY columnRefOrder ( COMMA columnRefOrder )* -> ^( TOK_ORDERBY ( columnRefOrder )+ ) );
+	public final HiveParser_IdentifiersParser.orderByClause_return orderByClause() throws RecognitionException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 130. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.